plugins/airbrake/src/api/AirbrakeApi.ts

```diff
@@ -24,3 +24,9 @@ export const airbrakeApiRef = createApiRef<AirbrakeApi>({
 export interface AirbrakeApi {
   fetchGroups(projectId: string): Promise<Groups>;
 }
+
+export class NoProjectIdError extends Error {
```

Comment:
This often doesn't work as expected, see https://github.com/backstage/backstage/blob/bc0b1f25c6ea4c5ab1d488b6ed7d8e1d96d6a606/packages/errors/src/errors/CustomErrorBase.ts#L54

Not sure about the approach of a custom error here tbh. What do you think about checking whether the `projectId` is available before calling the method instead?

Reply:
@Rugvip good shout. I'm keeping this custom error to stick with DRY principles with both the mock and production API. But I've removed conditionals checking the error type to be this custom error and instead just check for `projectId`.

As a result, I've managed to remove some more lines of code through further refactoring 😄 

Let me know what you think!

Reply:
Sorry not quite following. What I'm proposing is that you do something like this:

```ts
const { loading, value, error } = useAsync(
  () => projectId && airbrakeApi.fetchGroups(projectId),
  [airbrakeApi, projectId],
);
```

And then if `!projectId` you just render `<MissingAnnotationEmptyState />` straight away

Reply:
So I originally refactored this code because abruptly rendering a component in the logic felt hard to follow.

This is why I introduced the enums here as those are the 4 states that this widget can be in:
https://github.com/backstage/backstage/blob/4239b8dba6c490bc7cc97722e43c9699e3d849bf/plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx#L39-L44

`useAsync` requires me to either return the object from Airbrake or throw an error. I can't return undefined or an empty object (and don't want to make the type support these either). So I believe it does almost what you've proposed except it always either returns a result from Airbrake or throws an error.

Because it always has to either return a result or throw an error, having a conditional check for `projectId` won't have much value, because if `projectId` does not exist I still have to return or throw something.

So what I do is use the logic here:
https://github.com/backstage/backstage/blob/4239b8dba6c490bc7cc97722e43c9699e3d849bf/plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx#L62-L67

To set the `componentState` to the correct one in this case there not being a project ID.

I then still throw an error to satisfy `useAsync`:

https://github.com/backstage/backstage/blob/4239b8dba6c490bc7cc97722e43c9699e3d849bf/plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx#L68

And then print `<MissingAnnotationEmptyState />`:

https://github.com/backstage/backstage/blob/4239b8dba6c490bc7cc97722e43c9699e3d849bf/plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx#L95-L100

Hope that makes more sense :)